templates/install_puppet.ps1.erb

```diff
@@ -175,6 +175,66 @@ function Script:Unlock-Installation {
   }
 }
 
+<#
+.Synopsis
+  Collect the current state of puppet-agent services
+.Parameter service_names
+  an array of service names to save the state for
+#>
+function Script:Read-PuppetServices {
+  param(
+    [Parameter(Mandatory=$true)]
+    [Array] $service_names
+  )
+  begin{
+    $services = @()
+    foreach ($service_name in $service_names) {
+      Write-Log "Fetching current service status for $service_name"
+      # Let Get-Service silently continue if the service doesn't exist. This is
+      # so we can ignore mcollective when it doesn't exist (i.e. agent > 6.0.0)
+      $service_entry = Get-Service $service_name -ErrorAction SilentlyContinue
+      # The StartType entry was only added to 'Get-Service' in powershell 4, so
+      # we need to use WMI to fetch it for older versions of windows
+      if ($service_entry) {
+        $start_type = (Get-WmiObject -Class Win32_Service -Property StartMode -Filter "Name='$service_name'").StartMode.Replace('Auto', 'Automatic')
+        $services += @{ 'Name' = $service_name; 'Status' = $service_entry.Status; 'StartType' = $start_type }
+      } else {
+        Write-Log "Service $service_name does not exist, continuing..."
+      }
+    }
+    return $services
+  }
+}
+
+<#
+.Synopsis
+  Restore the state of puppet-agent services
+.Parameter services
+  An array containing hashtables of service statuses and start types in the form:
+  {'Name' = 'service name'; 'Status' = 'service status'; 'StartType' = 'service start type'}
+#>
+function Script:Reset-PuppetServices {
+  param(
+    [Parameter(Mandatory=$true)]
+    [AllowNull()]
+    [Array] $services
+  )
+  begin{
+    foreach ($service in $services) {
+      # We need to check if the service still exists after upgrade, since
+      # an upgrade from agent < 6 to agent >= 6 will have an mcollective
+      # service before upgrade, but no mcollective service after
+      if (Get-Service $service.Name -ErrorAction SilentlyContinue) {
+        Write-Log "Restoring service state for $($service.Name)"
+        Set-Service $service.Name -StartupType $service.StartType
+        Set-Service $service.Name -Status $service.Status
+      } else {
+        Write-Log "Skip $($service.Name), continuing..."

```

Comment:
May be better to just not log anything if a service no longer exists after the upgrade. Or at least indicate that's what's going on. "Skip" is a little vague.